benchmarks/distributed/rpc/ps/benchmark.py

```diff
@@ -0,0 +1,944 @@
+                        bp_loc_avg = bp_loc_avg.coalesce()
+                    bp_loc_avg = [
+                        bp_loc_avg._indices(),
+                        bp_loc_avg._values(),
+                        bp_loc_avg.size()
+                    ]
+                for cur_fut in self.futures[bp_loc]:
+                    cur_fut.set_result(bp_loc_avg)
+                record_event(
+                    self.rank,
+                    GRADIENT_SERVER_BATCH_METRIC,
+                    self.average_gradient_bp_key(
+                        self.batch_number
+                    ),
+                    BP_LOC_BATCH,
+                    self.metrics
+                )
+                return fut
+        return fut
+
+    def rpc_warmup_call(tensor):
+        return tensor
+
+    def reset(gs_rref):
+        self = gs_rref.local_value()
+        self.clear_batch_state()
+        self.batch_number = 0
+
+    def get_metrics(gs_rref):
+        self = gs_rref.local_value()
+        torch.cuda.synchronize(self.rank)
+        return [self.rank, process_metrics_for_elapsed_time(self.metrics)]
+
+
+class HookState():
+    def __init__(
+        self,
+        rank,
+        process_group,
+        process_group_size,
+        use_cuda_rpc,
+        gs_rref,
+        batch_number,
+        bp_location,
+        hook_gradient_futures,
+        metrics
+    ):
+        self.rank = rank
+        self.process_group = process_group
+        self.process_group_size = process_group_size
+        self.use_cuda_rpc = use_cuda_rpc
+        self.gs_rref = gs_rref
+        self.batch_number = batch_number
+        self.bp_location = bp_location
+        self.hook_gradient_futures = hook_gradient_futures
+        self.metrics = metrics
+
+    def next_batch_state(self):
+        self.bp_location = 0
+        self.batch_number += 1
+        self.hook_gradient_futures = []
+
+
+def hook_metric_key(state):
+    return "{},{}".format(state.batch_number, state.bp_location)
+
+
+def get_tensors(bucket):
+    parameter_tensors = bucket.get_per_parameter_tensors()
+    parameter_tensors_count = len(parameter_tensors)
+    if parameter_tensors_count > 0:
+        return parameter_tensors
+    else:
+        return [bucket.get_tensor()]
+
+
+def send_request_rpc(state, tensor, rpc_type, kwargs):
+    tensor = tensor.cuda(state.rank) if state.use_cuda_rpc else tensor.cpu()
+    record_event(
+        state.rank,
+        HOOK_METRIC,
+        hook_metric_key(state),
+        rpc_type,
+        state.metrics
+    )
+    fut = rpc.rpc_async(
+        state.gs_rref.owner(),
+        GradientServer.average_gradient,
+        args=(
+            state.gs_rref,
+            state.rank,
+            state.batch_number,
+            state.bp_location
+        ),
+        kwargs=kwargs
+    )
+    record_event(
+        state.rank,
+        HOOK_METRIC,
+        hook_metric_key(state),
+        rpc_type,
+        state.metrics
+    )
+    return fut
```

Comment:
just wanna make sure I understand this correctly. This metrics is not measure the RPC delay, right? Because the async RPC is not done yet. It's done when the future is marked as completed.

Reply:
okay, that was a misunderstanding on my part. I will update the end locations for future metrics.